### ISSUE TITLE
[rush] New community plugin Rush x Skott

### ISSUE THREAD

**Issue by antoine-coulon** [microsoft/rushstack#3945]
Hello rushers,

Lately I have been working on [skott](https://github.com/antoine-coulon/skott) which is a tool building graphs from an arbitrary project (mostly JavaScript/TypeScript stuff at the moment). Because I'm also using/loving Rush a lot, I started the [Krush project](https://github.com/antoine-coulon/krush) which will essentially be exposing a set of Rush plugins, _that will make you love Rush even more_ 😆 . I started by implementing a first plugin `rush-plugin-skott` which would allow to leverage some of the features of skott but adapted to Rush. It's just getting started, currently for the `v0.1.0` just release there is:

#### `rush graph`

This command generates a graph of your Rush monorepo projects and opens a webui to visualize it.

![skott-rush](https://user-images.githubusercontent.com/43391199/215892215-eee3b70a-4186-4669-b02c-18b0f376234e.png)

#### `rush list-unused-packages`

This command traverses each of the Rush monorepo projects and search for unused `npm` packages in production code.

![rush-list-unused](https://user-images.githubusercontent.com/43391199/216106483-4d6be883-1d2d-4bb0-a1ca-bc2c04884af6.png)

Everything could be adapted specifically to fit Rush's needs.

So now two questions:

- Is this possible to feature that in the [`Third-party plugins section`](https://rushjs.io/pages/maintainer/using_rush_plugins/)? I can land a PR wherever this needs to be done

- What features/requirements would need to be achieved to integrate some of these plugin commands directly in Rush as builtin commands? If you guys have any input or any suggestion about some features/things that would make sense, we can discuss about it. I'm not sure Rush (or third-party plugin) has already any _graph visualizer_ yet nor things like _unused dependencies tracker_. As far as I'm concerned, the `list-unused-packages` command could be interactive and suggest like the `upgrade-interactive` which dependencies to prune.

Moreover, these commands don't take any CLI arguments but this could be easily done (filter projects, etc).

Thanks! 🏍️

**Comment by iclanton** [comment]
>Is this possible to feature that in the [Third-party plugins section](https://rushjs.io/pages/maintainer/using_rush_plugins/)? I can land a PR wherever this needs to be done

Absolutely.

>What features/requirements would need to be achieved to integrate some of these plugin commands directly in Rush as builtin commands? If you guys have any input or any suggestion about some features/things that would make sense, we can discuss about it. I'm not sure Rush (or third-party plugin) has already any graph visualizer yet nor things like unused dependencies tracker. As far as I'm concerned, the list-unused-packages command could be interactive and suggest like the upgrade-interactive which dependencies to prune.

We don't have a graph visualizer built-in yet. I'd be open to build it in directly to Rush once the plugin looks stable, depending on how large its dependency footprint ends up being.

WRT `rush list-unused-packages`, the one of libraries that @TheLarkInn used for `rush upgrade-interactive` (`npm-check`) already has some of this functionality built-in, but I don't know how well it actually works. We have an internal `heft` plugin that scans for unused dependencies during build, but I think it also makes sense for Rush to have this functionality. Pulling this directly into Rush will also depend on its dependency footprint.

>Moreover, these commands don't take any CLI arguments but this could be easily done (filter projects, etc).

I think it'd be useful to support the project selectors for these arguments (i.e. - `--from`, `--to`, `--only`, `--to-except`, etc.).

**Comment by iclanton** [comment]
@antoine-coulon - Would you be interested in demoing this at a Rush Hour event (https://rushstack.io/community/events/). The next one is this upcoming Friday.

**Comment by antoine-coulon** [comment]
> @antoine-coulon - Would you be interested in demoing this at a Rush Hour event (https://rushstack.io/community/events/). The next one is this upcoming Friday.

Yeah for sure I'd be pleased to join! Could you please share some information about the format of the meeting so that I can prepare a small demo around this graphic visualization?

**Comment by octogonz** [comment]
> Yeah for sure I'd be pleased to join! Could you please share some information about the format of the meeting so that I can prepare a small demo around this graphic visualization?

Hi @antoine-coulon,

The meeting is 60 minutes and separated into three parts, each about 20 mins in length:

- **roadmap:** We review interesting PRs merged last month, and the plan for features coming soon
- **segments:** People share short demos, tech sharing presentations, or design proposals. (These are prepared in advance.)
- **Q&A:** The remainder of the meeting is open discussion for tech support questions, suggesting new features, or any other topic of interest

Your demo can be very short (just a couple minutes) but generally not longer than 15 mins.  It is a Microsoft Teams video call, so you can share your screen or slide deck.

If you have questions feel free to DM me on [Zulip](https://rushstack.zulipchat.com/#narrow/pm-with/357793-user357793) (using your GitHub account), [Twitter](https://twitter.com/octogonz_), or [Mastodon](https://mastodon.social/@octogonz).  Thanks!

**Comment by antoine-coulon** [comment]
Hey @octogonz thanks for all the detailed information. See you on friday!

**Comment by octogonz** [comment]
@antoine-coulon Thanks for sharing at Rush Hour today! 🏆 Great demo!

We're looking forward to seeing how this plugin evolves. I will make sure it's featured in the plugins gallery on the Rush website.

**Comment by antoine-coulon** [comment]
@octogonz thanks for having me it was a great experience 😄

I'm looking forward to make this plugin evolve in the next few weeks taking into account all the feedbacks I received.

Should we let this issue open to exchange about this subject?

By the way I just joined Zulipe, I'll share my slides on it as promised.

See you next Rush Hour (West) hopefully!

**Comment by octogonz** [comment]
> Should we let this issue open to exchange about this subject?

Yes, we sometimes keep a GitHub issue open for general discussion/feedback about a topic.

**Comment by antoine-coulon** [comment]
Hello guys, hope you're doing well. I didn't have time to move forward on this subject yet. `skott` itself is evolving quickly and the Rush plugin will for sure benefit from that in the next few weeks. I'll keep you updated
